feat: add batch (#256)
near/NEPs: specs/Standards/NonFungibleToken/Event.md

This extension addresses that.

For example, it's common in NFT marketplaces to have
different methods of transfer a single token vs many 
tokens. Other applications like wallets, will have
a difficult time tracking consistently this information for many
markets. This makes interactive with many NFT-driven apps 
infeasible.

Due to discussions here 
https://github.com/near/NEPs/issues/254,
a standard way to capture these events is needed.

## Interface

Many apps may use different mechanisms to perform their batching
requirements. This interface standardizes this process by capturing
events of these activities through logs.

```ts
// Interface to capture an event
// and return formatted event string.
interface EventJsonLog {
    // Takes `EventLogData` and returns
    // `EVENT_JSON: <EVENT_LOG_DATA>`
    (e:EventLogData):string
}

// Interface to capture data 
// about an event
// Arguments
// * `standard`: name of standard e.g. nep171
// * `version`: e.g. 1.0.0
// * `event`: `nft_mint` | `nft_burn` | `nft_transfer`
// * `data`: associate event data
interface EventLogData {
    standard:string,
    version:string,
    event:string,

>     data: NftMintLog[]|NftTransferLog[]|NftBurnLog

Reviewer: https://github.com/near/NEPs/pull/256/files#r726534183

}

// An event log to capture token minting
// Arguments
// * `account_id`: "account.near"
// * `token_id`: "1"
interface NftMintLog {
    owner_id:string,

>     token_id:string

Reviewer 2: We need to add an optional `memo` field here and to all other event log types.

Author: What does `memo` represent?

I'm not aware of any `nft_mint` function that uses memo. For example https://github.com/near/near-sdk-rs/blob/master/examples/non-fungible-token/nft/src/lib.rs#L92

}

// An event log to capture token burning
// Arguments

> // * `owner_id`: owner of tokens to burn

Author: Interesting - performing an action on behalf of the owner. Do we want this information in logs? Because all actions could be performed on behalf of owners and we'll need `approval_id` in all events. 

The more information we add, the more chance of hitting the 16kb limitation is. I based these events on how they are actually being used today:

https://github.com/zavodil/pluminite-ui/blob/master/contract/src/burn.rs#L7
https://github.com/near-apps/uhhm-market/blob/main/contracts/nft-simple/src/burn.rs#L7
https://github.com/ParasHQ/paras-nft-contract/blob/main/paras-nft-contract/src/lib.rs#L579

If this is not priority now, in the next version we can add this.

> // * `owner_id`: owner of tokens to burn

Reviewer: Agreed to not call it `approval_id` as that's a number and we're only wanting to emit info on the account that burned it. Still needs to add some name for that.
TODO good naming for this.

// * `token_ids`: ["1","2"]
interface NftBurnLog {
    owner_id:string,

>     token_ids:string[]

Reviewer 2: Are we sure we want to pass here the list? I think it's better to use here `token_id` as everywhere else not to mess up.

Reviewer: I think in the majority of cases for batch minting and burning, these will be from the same user account and this makes sense. Similar to suggestions I'm seeing in the [multi-token standard](https://github.com/near/NEPs/pull/245).

Reviewer 2: Do we want to pass the list of ids in mint also?

Not suggesting doing that for transfer, it sounds weird. But I didn't know the case with burn, so I could be wrong here.

Reviewer: Good point, I definitely think we should do that with mint

Reviewer 3: @evergreen-trading-system @telezhnaya @mikedotexe I believe that if we go with batching all the way down, we should bake this support everywhere: mint, transfer, burn.

```js
JSON_EVENT {"standard": "nep171", "version": "1.0.0", "event": "mint": "data": [
  {"owner_id": "frol.near", "token_ids": ["1", "2"]}
]}

JSON_EVENT {"standard": "nep171", "version": "1.0.0", "event": "transfer": "data": [
  {"sender_id": "frol.near", "receiver_id": "olga.near", "token_ids": ["1", "2"]}
]}

JSON_EVENT {"standard": "nep171", "version": "1.0.0", "event": "burn": "data": [
  {"owner_id": "frol.near", "approval_id": "olga.near", "token_ids": ["1", "2"]},
  {"owner_id": "frol.near", "token_ids": ["3"]}
]}
```

Reviewer 3: @mikedotexe Do you have a good naming for "an account that was approved to perform transfer/burn operations on behalf of the NFT owner"? I currently named it as "representative_id" just for the sake of no better name I could come up with. I feel it is similar to "Intermediate / Corresponding Bank", but I am really bad in that area of terminology.

}

// An event log to capture token transfer
// Arguments
// * `sender_id`: "account.near"
// * `receiver_id`: "receiver.near"
// * `token_id`: "12345abc"
interface NftTransferLog {
    sender_id:string,
    receiver_id:string,
    token_id:string
}
```

## Further methods

Note that the example events covered above cover two different kinds of events:
1. Events that are not specified in the NFT Standard (`nft_mint`, `nft_burn`)

> 2. An event that is covered in the [NFT Core Standard](https://nomicon.io/Standards/NonFungibleToken/Core.html#nft-interface).

Reviewer: https://github.com/near/NEPs/pull/256/files#r726530405

>  This event standard also applies beyond the three events highlighted here, where future events follow the same convention of as the second type. For instance, if an NFT contract uses the [approval management standard](https://nomicon.io/Standards/NonFungibleToken/ApprovalManagement.html), it may emit an event for `nft_approve` if that's deemed as important by the developer community.

Author: So going to make separate pr for approval events

Each app may have their own way of performing these actions.
This introduces inconsistency in capturing these events.
This extension addresses that.

It's common for NFT apps to transfer many or one token at a time.
Other applications need to track these and similar events consistently.
If not, tracking state across many NFT-driven apps become infeasible.

We need a standard way to capture events.
This was discussed here https://github.com/near/NEPs/issues/254.

## Interface

Many apps use different interfaces that represent the same action.
This interface standardizes that process.
It captures these actions through logs.

```ts
// Interface to capture an event
// and return formatted event string.
interface EventJsonLog {
    // Takes `EventLogData` and returns
    // `EVENT_JSON: <EVENT_LOG_DATA>`
    (e:EventLogData):string
}

// Interface to capture data 
// about an event
// Arguments
// * `standard`: name of standard e.g. nep171
// * `version`: e.g. 1.0.0
// * `event`: `nft_mint` | `nft_burn` | `nft_transfer`
// * `memo`: optional message
// * `data`: associate event data
interface EventLogData {
    standard:string,
    version:string,
    event:string,

>     memo?:string,

Reviewer 2: ```suggestion
```

# Non-Fungible Token Event

Version `1.0.0`

## Summary

Standard interfaces for NFT contract actions.

## Motivation

NFT-driven apps perform many similar actions.
For example - `minting`, `burning` and `transferring`.
Each app may have their own way of performing these actions.
This introduces inconsistency in capturing these events.
This extension addresses that.

It's common for NFT apps to transfer many or one token at a time.
Other applications need to track these and similar events consistently.
If not, tracking state across many NFT-driven apps become infeasible.

We need a standard way to capture events.
This was discussed here https://github.com/near/NEPs/issues/254.

## Interface

Many apps use different interfaces that represent the same action.
This interface standardizes that process.
It captures these actions through logs.

```ts
// Interface to capture an event
// and return formatted event string.
interface EventJsonLog {
    // Takes `EventLogData` and returns
    // `EVENT_JSON: <EVENT_LOG_DATA>`

>     (e:EventLogData):string

Reviewer 2: Are we sure that we don't want to put everything into json with the structure like
```
{
"type": "event_json",
"data": [...] // here could be all events
}
```
?

It simplifies everything, we don't need to overcomplicate this doc, Indexer does not need the logic with cutting/trimming.

>     (e:EventLogData):string

Author: @mikedotexe

>     (e:EventLogData):string

Reviewer 3: @telezhnaya Individual calls to `near_sdk::env::log` are recorded as individual strings in the `logs` array of the execution outcome, so we don't care about new lines inside. We only want to specify that each event should be a separate call to the log host function.

Proof: https://explorer.testnet.near.org/transactions/326HmzvgRYND7skVCdTpCnHMji6qiU8JFhQMNuvS2b74
![image](https://user-images.githubusercontent.com/304265/137962691-0eb3aa1f-c228-4e9b-a73e-447f96efd8be.png)

Contract:

```rust
use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
use near_sdk::{near_bindgen, setup_alloc};

setup_alloc!();

#[near_bindgen]
#[derive(BorshDeserialize, BorshSerialize, Default)]
pub struct Test {}

#[near_bindgen]
impl Test {
    pub fn test_log(&mut self) {
        near_sdk::env::log(r#"EVENT_JSON:{
            "standard": "nep",
            "version": "1.0.0",
            "event": "mint",
            "data": [
                {"owner_id": "frol.near", "token_ids": ["house"]}
            ]
        }"#.as_bytes());
        near_sdk::env::log("EVENT_JSON:{\"standard\": \"nep\", \n \"version\": \"1.0.0\",    \n   \"event\": \"burn\", \"data\": [{\"owner_id\": \"frol.near\", \"token_ids\": [\"house\"]}]}".as_bytes());
    }
}
```

// about an event
// Arguments
// * `standard`: name of standard e.g. nep171
// * `version`: e.g. 1.0.0
// * `event`: `nft_mint` | `nft_burn` | `nft_transfer`
// * `data`: associate event data
interface EventLogData {
    standard:string,
    version:string,
    event:string,
    data: NftMintLog[]|NftTransferLog[]|NftBurnLog[]
}

// An event log to capture token minting
// Arguments
// * `owner_id`: "account.near"
// * `token_ids`: ["1", "abc"]
// * `memo`: optional message
interface NftMintLog {
    owner_id:string,
    token_ids:string[],
    memo?:string
}

// An event log to capture token burning
// Arguments
// * `owner_id`: owner of tokens to burn
// * `authorised_id`: approved account to burn
// * `token_ids`: ["1","2"]
// * `memo`: optional message
interface NftBurnLog {
    owner_id:string,
    authorised_id:string,
    token_ids:string[],
    memo?:string
}

// An event log to capture token transfer
// Arguments
// * `authorised_id`: approved account to transfer
// * `sender_id`: "account.near"
// * `receiver_id`: "receiver.near"
// * `token_ids`: ["1", "12345abc"]
// * `memo`: optional message
interface NftTransferLog {
    authorised_id:string,

>     sender_id:string,

Reviewer 3: I think @telezhnaya suggested:

```js
{
owner_id:"owner.near",
authorised_id:"authorised.sender.near",
receiver_id:"receiver.near",
token_ids: ["1"]
}
```

Though, if we change sender to owner we get a different sort of confusion, when `authorised_id` is omitted:

```js
{
owner_id:"owner.near",
receiver_id:"receiver.near",
token_ids: ["1"]
}
```

I don't have a suggestion so far, but I agree with (sender, authorised, receiver) being confusing since sender is not actually sending in this case, and (owner, receiver) is not immediately clear as well.

>     sender_id:string,

Author: ```ts
{
old_owner_id:"owner.near",
authorised_id:"authorised.sender.near",
new_owner_id:"receiver.near",
token_ids: ["1"]
}
```

Makes sense. Though it won't have symmetrical interface with others. Is this ok?

>     sender_id:string,

Author: Can't answer on how `authorised_id` will be used but I can say all useful events in general e.g. `ApprovalEvent` you mentioned is part of the plan - standard nft events. 

I will be creating another pr shortly regarding and I invite to mention as many events you think should be standardised.

>     sender_id:string,

Reviewer 4: thanks @frol makes sense!

Standard interfaces for NFT contract actions.

## Motivation

NFT-driven apps perform many similar actions.
For example - `minting`, `burning` and `transferring`.
Each app may have their own way of performing these actions.
This introduces inconsistency in capturing these events.
This extension addresses that.

It's common for NFT apps to transfer many or one token at a time.
Other applications need to track these and similar events consistently.
If not, tracking state across many NFT-driven apps become infeasible.

We need a standard way to capture events.
This was discussed here https://github.com/near/NEPs/issues/254.

## Interface

Many apps use different interfaces that represent the same action.
This interface standardizes that process.
It captures these actions through logs.

```ts
// Interface to capture an event
// and return formatted event string.
interface EventJsonLog {
    // Takes `EventLogData` and returns
    // `EVENT_JSON: <EVENT_LOG_DATA>`
    (e:EventLogData):string
}

// Interface to capture data 
// about an event
// Arguments
// * `standard`: name of standard e.g. nep171
// * `version`: e.g. 1.0.0
// * `event`: `nft_mint` | `nft_burn` | `nft_transfer`

> // * `memo`: optional message

Reviewer: ```suggestion
```
Think this was an accidental copy/paste

```

#### Invalid event logs:

* Two events in a single log entry (instead, call `log` for each individual event)
```
EVENT_JSON:{"standard": "nepXXX", "version": "1.0.0", "event": "xyz_is_triggered"}
EVENT_JSON:{"standard": "nepXXX", "version": "1.0.0", "event": "xyz_is_triggered"}
```
* Invalid JSON data
```
EVENT_JSON:invalid json
```

## Interface

Non-Fungible Token Events MUST have `standard` set to `"nep171"`, standard version set to `"1.0.0"`, `event` value is one of `nft_mint`, `nft_burn`, `nft_transfer`, and `data` must be of one of the following relavant types: `NftMintLog[] | NftTransferLog[] | NftBurnLog[]`:

```ts
interface EventLogData {
    standard: "nep171",
    version: "1.0.0",
    event: "nft_mint" | "nft_burn" | "nft_transfer",
    data: NftMintLog[] | NftTransferLog[] | NftBurnLog[],
}
```

```ts
// An event log to capture token minting
// Arguments
// * `owner_id`: "account.near"
// * `token_ids`: ["1", "abc"]
// * `memo`: optional message
interface NftMintLog {
    owner_id: string,
    token_ids: string[],
    memo?: string
}

// An event log to capture token burning
// Arguments
// * `owner_id`: owner of tokens to burn

> // * `authorised_id`: approved account to burn, if applicable

Reviewer: ```suggestion
// * `authorized_id`: approved account to burn, if applicable
```
Hate to be nit-picky about this, but I do see other references in the NEP to the word "authorize" and "authorized" and think this will be an introduction of a word spelling that is inconsistent with numerous cases in our docs, NEAR CLI, nearcore, Rust SDK, and other NEPs.

So I think we'll have to stick with the spelling we have in these other places for consistently, lest we cause confusion for end users.

<img width="1764" alt="Screen Shot 2021-10-21 at 2 25 37 PM" src="https://user-images.githubusercontent.com/1042667/138359406-82b1de93-589e-4897-9e7b-c820e38ce890.png">
<img width="944" alt="Screen Shot 2021-10-21 at 2 23 23 PM" src="https://user-images.githubusercontent.com/1042667/138359424-3a6ddf20-a4fa-4694-b7ac-adaa1c9b3385.png">
<img width="1359" alt="Screen Shot 2021-10-21 at 2 22 05 PM" src="https://user-images.githubusercontent.com/1042667/138359426-557f2c27-206a-483f-a521-03a62918a226.png">
![Screen Shot 2021-10-21 at 2 20 42 PM](https://user-images.githubusercontent.com/1042667/138359429-a5ea33e1-474a-4198-8546-0478dd07ef5b.png)
<img width="1322" alt="Screen Shot 2021-10-21 at 2 19 42 PM" src="https://user-images.githubusercontent.com/1042667/138359430-c540c6f0-9b01-41c7-b9f7-ef54d4d7d92c.png">

// * `token_ids`: ["1","2"]
// * `memo`: optional message
interface NftBurnLog {
    owner_id: string,

>     authorised_id?: string,

Reviewer: ```suggestion
    authorized_id?: string,
```

    token_ids: string[],
    memo?: string
}

// An event log to capture token transfer
// Arguments

> // * `authorised_id`: approved account to transfer

Reviewer: ```suggestion
// * `authorized_id`: approved account to transfer
```

// * `old_owner_id`: "owner.near"
// * `new_owner_id`: "receiver.near"
// * `token_ids`: ["1", "12345abc"]
// * `memo`: optional message
interface NftTransferLog {

>     authorised_id?: string,

Reviewer: ```suggestion
    authorized_id?: string,
```

# Non-Fungible Token Event

Version `1.0.0`

## Summary

Standard interfaces for NFT contract actions.

## Motivation

> Nft driven apps such as marketplaces regularly perform many

Reviewer: ```suggestion
NFT-driven apps such as marketplaces regularly perform many
```

> of the same acions such as `minting`, `burning` and `transfering`.

Reviewer: ```suggestion
of the same actions such as `minting`, `burning` and `transferring`.
```

Each app have their own way of performing these actions and it
is difficult to consistently capture these events.
This extension addresses that.

For example, it's common in NFT marketplaces to have
different methods of transfer a single token vs many 
tokens. Other applications like wallets, will have
a difficult time tracking consistently this information for many

> markets. This makes interactive with many nft driven apps 

Reviewer: ```suggestion
markets. This makes interactive with many NFT-driven apps 
```

infeasible.

Due to discussions here 
https://github.com/near/NEPs/issues/254,

> a standard way to capture these events are needed.

Reviewer: ```suggestion
a standard way to capture these events is needed.
```

## Interface

Many apps may use different mechanisms to perform their batching

> requirements. This interface standardises this process by capturing

Reviewer: ```suggestion
requirements. This interface standardizes this process by capturing
```

events of these activities through logs.

```ts
// Interface to capture an event
// and return formatted event string.
interface EventJsonLog {
    // Takes `EventLogData` and returns
    // `EVENT_JSON: <EVENT_LOG_DATA>`
    (e:EventLogData):string

> }

Author: This notifies implementors of their need to produce the final string compatible with the indexer. Look at line 39 `EVENT_JSON {"standard":"nep171", ...}`.

This is the response to https://github.com/near/NEPs/issues/254#issuecomment-928249174 to capture all kinds of events not just json.

// Interface to capture data 
// about an event
// Arguments
// * `standard`: name of standard e.g. nep171
// * `version`: e.g. 1.0.0
// * `event`: `nft_mint` | `nft_burn` | `nft_transfer`
// * `data`: associate event data
interface EventLogData {
    standard:string,
    version:string,
    event:string,
    data: NftMintLog[]|NftTransferLog[]|NftBurnLog
}

// An event log to capture token minting
// Arguments
// * `account_id`: "account.near"
// * `token_id`: "1"
interface NftMintLog {
    owner_id:string,
    token_id:string
}

// An event log to capture token burning
// Arguments
// * `owner_id`: owner of tokens to burn
// * `token_ids`: ["1","2"]
interface NftBurnLog {
    owner_id:string,
    token_ids:string[]
}

// An event log to capture token transfer
// Arguments
// * `sender_id`: "account.near"
// * `receiver_id`: "receiver.near"
// * `token_id`: "12345abc"
interface NftTransferLog {
    sender_id:string,
    receiver_id:string,
    token_id:string
}
```

> 

Reviewer: ```suggestion

## Further methods

Note that the example events covered above cover two different kinds of events:
1. Events that are not specified in the NFT Standard (`nft_mint`, `nft_burn`)
2. An event that is covered in the [NFT Core Standard](https://nomicon.io/Standards/NonFungibleToken/Core.html#nft-interface).

 This event standard also applies beyond the three events highlighted here, where future events follow the same convention of as the second type. For instance, if an NFT contract uses the [approval management standard](https://nomicon.io/Standards/NonFungibleToken/ApprovalManagement.html), it may emit an event for `nft_approve` if that's deemed as important by the developer community.
 
 Please feel free to open pull requests for extending the events standard detailed here as needs arise.
```
